csharp/CompoundLogicalColumnWriter.cs

```diff
@@ -0,0 +1,322 @@
+            }
+
+            if (IsNested(elementType, out var innerNestedRequired))
+            {
+                if (schemaNodes.Length >= 1 &&
+                    schemaNodes[0] is GroupNode { LogicalType: NoneLogicalType, Repetition: Repetition.Required })
+                {
+                    return MakeGenericWriter(nameof(MakeNestedWriter), innerNestedRequired, schemaNodes.Skip(1).ToArray(),
+                        repetitionLevel, nullDefinitionLevel, firstLeafRepLevel);
+                }
+
+                throw new Exception("elementType is nested (required) but schema does not match expected layout");
+            }
+
+            if (elementType.IsArray && elementType != typeof(byte[]))
+            {
+                if (schemaNodes.Length >= 2 &&
+                    schemaNodes[0] is GroupNode { LogicalType: ListLogicalType, Repetition: Repetition.Optional } &&
+                    schemaNodes[1] is GroupNode { LogicalType: NoneLogicalType, Repetition: Repetition.Repeated })
+                {
+                    var containedType = elementType.GetElementType();
+
+                    return MakeArrayWriter(
+                        schemaNodes.Skip(2).ToArray(),
+                        containedType,
+                        nullDefinitionLevel,
+                        repetitionLevel,
+                        firstLeafRepLevel
+                    );
+                }
+
+                throw new Exception("elementType is an array but schema does not match the expected layout");
+            }
+
+            if (schemaNodes.Length == 1)
+            {
+                bool isOptional = schemaNodes[0].Repetition == Repetition.Optional;
+
+                short leafDefinitionLevel = isOptional ? (short)(nullDefinitionLevel + 1) : nullDefinitionLevel;
+                short leafNullDefinitionLevel = isOptional ? nullDefinitionLevel : (short)-1;
+
+                if (singleItem)
+                {
+                    return MakeLeafWriterSingle(repetitionLevel, firstLeafRepLevel, leafDefinitionLevel, leafNullDefinitionLevel);
+                }
+                else
+                {
+                    return MakeLeafWriter(repetitionLevel, firstLeafRepLevel, leafDefinitionLevel, leafNullDefinitionLevel);
+                }
+            }
+
+            throw new Exception("ParquetSharp does not understand the schema used");
+        }
+
+        private Action<Array> MakeArrayWriter(Node[] schemaNodes, Type elementType, short nullDefinitionLevel, short repetitionLevel, short firstLeafRepLevel)
+        {
+            var columnWriter = (ColumnWriter<TPhysical>)Source;
+
+            var writer0 = MakeWriter(schemaNodes, elementType, (short)(repetitionLevel + 1), (short)(nullDefinitionLevel + 2), firstLeafRepLevel, false);
+            var writer = MakeWriter(schemaNodes, elementType, (short)(repetitionLevel + 1), (short)(nullDefinitionLevel + 2), repetitionLevel, false);
```

Comment:
It took me a while to wrap my head around why we need a `writer0` and `writer` with different `firstLeafRepLevel`, would be good to try to have a thorough comment explaining this part.